Core/include/Acts/Utilities/Any.hpp

```diff
@@ -0,0 +1,427 @@
+    if (m_handler == nullptr) {
+      return;
+    }
+
+    void* to = m_data.data();
+    const void* from = fromAny.m_data.data();
+    if (m_handler->typeSize > SIZE) {
+      // stored on heap: interpret buffer as pointer
+      to = *reinterpret_cast<void**>(m_data.data());
+      from = *reinterpret_cast<void* const*>(fromAny.m_data.data());
+    }
+
+    if (m_handler->copyConstruct == nullptr) {
+      // trivially copyable
+      m_data = fromAny.m_data;
+    } else {
+      m_handler->copyConstruct(from, to);
+    }
+  }
+
+  void copy(const AnyBase& fromAny) {
+    if (m_handler == nullptr) {
+      return;
+    }
+
+    void* to = m_data.data();
+    const void* from = fromAny.m_data.data();
+    if (m_handler->typeSize > SIZE) {
+      // stored on heap: interpret buffer as pointer
+      to = *reinterpret_cast<void**>(m_data.data());
+      from = *reinterpret_cast<void* const*>(fromAny.m_data.data());
+    }
+
+    if (m_handler->copy == nullptr) {
+      // trivially copyable
+      m_data = fromAny.m_data;
+    } else {
+      m_handler->copy(from, to);
+    }
+  }
+
+  template <typename T>
+  static void destroyImpl(void* ptr) {
+    assert(ptr != nullptr && "Address to destroy is nullptr");
+    T* obj = static_cast<T*>(ptr);
+    if constexpr (sizeof(T) <= SIZE) {
+      // stored in place: just call the destructor
+      _ACTS_ANY_DEBUG("Destroy local at: " << ptr);
+      obj->~T();
+    } else {
+      // stored on heap: delete
+      _ACTS_ANY_DEBUG("Delete heap at: " << obj);
+      delete obj;
+    }
+  }
+
+  template <typename T>
+  static void moveConstructImpl(void* from, void* to) {
+    _ACTS_ANY_DEBUG("move const: " << from << " -> " << to);
+    assert(from != nullptr && "Source is null");
+    assert(to != nullptr && "Target is null");
+    T* _from = static_cast<T*>(from);
+    /*T* ptr =*/new (to) T(std::move(*_from));
+  }
+
+  template <typename T>
+  static void moveImpl(void* from, void* to) {
+    _ACTS_ANY_DEBUG("move: " << from << " -> " << to);
+    assert(from != nullptr && "Source is null");
+    assert(to != nullptr && "Target is null");
+
+    T* _from = static_cast<T*>(from);
+    T* _to = static_cast<T*>(to);
+
+    (*_to) = std::move(*_from);
+  }
+
+  template <typename T>
+  static void copyConstructImpl(const void* from, void* to) {
+    _ACTS_ANY_DEBUG("copy const: " << from << " -> " << to);
+    assert(from != nullptr && "Source is null");
+    assert(to != nullptr && "Target is null");
+    const T* _from = static_cast<const T*>(from);
+    /*T* ptr =*/new (to) T(*_from);
+  }
+
+  template <typename T>
+  static void copyImpl(const void* from, void* to) {
+    _ACTS_ANY_DEBUG("copy: " << from << " -> " << to);
+    assert(from != nullptr && "Source is null");
+    assert(to != nullptr && "Target is null");
+
+    const T* _from = static_cast<const T*>(from);
+    T* _to = static_cast<T*>(to);
+
+    (*_to) = *_from;
+  }
+
+  alignas(std::max_align_t) std::array<char, SIZE> m_data{};
```

Comment:
Note that max_align_t is not enough for Eigen types, which may require AVX (32B) or AVX-512 (64B) SIMD alignment. It's only meant to provide the largest possible alignement for a scalar type :

> std::max_align_t is a [trivial](https://en.cppreference.com/w/cpp/named_req/TrivialType) [standard-layout](https://en.cppreference.com/w/cpp/named_req/StandardLayoutType) type whose [alignment requirement](https://en.cppreference.com/w/cpp/language/object#Alignment) is at least as strict (as large) as that of every scalar type.

Reply:
Is there a good alternative?

Reply:
Not in standard C/++ as far as I know. The C11 standard says...

> It is implementation-defined if over-aligned types are supported, and their support may be different in each kind of [storage duration](https://en.cppreference.com/w/c/language/storage_duration).

As a non-standard extension with a portable fallback, you could do...

```c++
static constexpr size_t alignment = 
#ifdef __x86_64__
    64  // Good enough for AVX-512
#else
    alignof(std::max_align_t)
#endif
;
```

Of course, this idea can be further refined by testing for presence of AVX and AVX-512:

```c++
static constexpr size_t alignment = 
#ifdef __AVX512F__
    64
#elif defined(__AVX__)
    32
#elif defined(__SSE__)
    16
#else
    alignof(std::max_align_t)
#endif
;
```

Reply:
Ok, changed to the fancy version from your comment.